grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy

```diff
@@ -162,31 +162,47 @@ class IndicatorDataService {
         return indicatorData;
     }
 
-    DataGraph getReceivedStockData(Location location) {
-        List listData = []
-        List listLabel = []
+    DataGraph getReceivedStockData(Location location, def params) {
+        Integer querySize = params.querySize? params.querySize.toInteger()-1 : 5
         today.clearTime()
-        for(int i=5;i>=0;i--){
-            def monthBegin = today.clone()
-            def monthEnd = today.clone()
-            monthBegin.set(month: today.month - i, date: 1)
-            monthEnd.set(month: today.month - i + 1, date: 1)
-                
-            def temp = Requisition.executeQuery("""select count(r) from Requisition r where r.dateCreated >= :monthOne and r.dateCreated < :monthTwo and r.destination = :location""",
-            ['monthOne': monthBegin, 'monthTwo': monthEnd, 'location': location]);
-            String monthLabel = new java.text.DateFormatSymbols().months[monthBegin.month]
+        
+        Date queryLimit = today.clone()
+        queryLimit.set(month: today.month - querySize, date: 1) 
 
-            listLabel.push(monthLabel)
-            listData.push(temp[0])
-        }
+        List queryData = Shipment.executeQuery("SELECT COUNT(s.id), s.origin, MONTH(s.lastUpdated), YEAR(s.lastUpdated) FROM Shipment s WHERE s.destination = :location AND s.currentStatus <> 'PENDING' AND s.lastUpdated > :limit GROUP BY MONTH(s.lastUpdated), YEAR(s.lastUpdated), s.origin", 
+        ['location': location, 'limit': queryLimit])
+        
+        List listRes = []
+        List listLabel = []
+        for(item in queryData) {
+            // item[0]: item total counted
+            // item[1]: item origin
+            // item[2]: item month
+            // item[3]: item year
+            Location itemLocation = item[1]
+            List listData = []
+            listLabel = []
 
-        List<IndicatorDatasets> datasets = [
-            new IndicatorDatasets('Stock movements received', listData)
-        ];
+            for(int i = querySize; i >= 0; i--) {
+                Date month = today.clone()
```

Comment:
Could you change this date name? Or even do you need a date obj for the month? Can't you use Integer? I think it would be better to avoid this type of code `months[month.month]` (it's ugly).

Reply:
When we subtract i months from the object, we also need to update Year when necessary: That's why we are using a date obj instead a simple Integer. 
(As we can filter data from more than a year ago, Year matters)
Therefore I'm going to just change this date's name.